Implement working Clone and PartialEq on DataPayload and Yoke

Currently, the `Clone` impl is present for `Yoke` and `DataPayload`, but it does not work when one tries to use it.  The error looks like this:

```
note: the following trait bounds were not satisfied:
      `<hello_world::HelloWorldV1<'static> as Yokeable<'a>>::Output: Clone`
```

This seems to be caused by https://github.com/rust-lang/rust/issues/85636.  However, the workaround mentioned there does not work for `Clone`, since the Clone impl for a reference copies the reference instead of cloning the struct that the reference points to.

`PartialEq` is a related but separate issue.  I have the following impl, but it doesn't compile:

```rust
impl<'d, 's, M> PartialEq for DataPayload<'d, 's, M>
where
    M: DataMarker<'s>,
    for<'a> &'a <M::Yokeable as Yokeable<'a>>::Output: PartialEq,
{
    fn eq(&self, other: &Self) -> bool {
        self.get().eq(other)
    }
}
```

Error message:

```
error[E0308]: mismatched types
   --> provider/core/src/data_provider.rs:165:23
    |
165 |         self.get().eq(other)
    |                       ^^^^^ expected reference, found struct `data_provider::DataPayload`
    |
    = note: expected reference `&&<<M as marker::DataMarker<'s>>::Yokeable as Yokeable<'_>>::Output`
               found reference `&data_provider::DataPayload<'d, 's, M>`
```

I played around with different variations of `&` and `*` on the arguments to get them to match up, to no success.  My guess is that the compiler is just confused about the lifetimes, since the lifetime of `.get()` is different than the lifetime of `other`.

CC @Manishearth

--- a/provider/core/src/data_provider.rs
+++ b/provider/core/src/data_provider.rs
@@ -10,6 +10,7 @@ use crate::error::Error;
 use crate::marker::DataMarker;
 use crate::resource::ResourceKey;
 use crate::resource::ResourcePath;
+use crate::yoke::trait_hack::YokeTraitHack;
 use crate::yoke::*;
 
 use alloc::rc::Rc;
@@ -142,10 +143,8 @@ where
 impl<'d, 's, M> Clone for DataPayload<'d, 's, M>
 where
     M: DataMarker<'s>,
-    for<'a> <M::Yokeable as Yokeable<'a>>::Output: Clone,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
 {
-    /// Note: This function is currently inoperable. For more details, see
-    /// https://github.com/unicode-org/icu4x/issues/753
     fn clone(&self) -> Self {
         use DataPayloadInner::*;
         let new_inner = match &self.inner {
@@ -156,6 +155,31 @@ where
         };
         Self { inner: new_inner }
     }
+}
+
+impl<'d, 's, M> PartialEq for DataPayload<'d, 's, M>
+where
+    M: DataMarker<'s>,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        YokeTraitHack(self.get()).into_ref() == YokeTraitHack(other.get()).into_ref()
+    }
+}
+
+impl<'d, 's, M> Eq for DataPayload<'d, 's, M>
+where
+    M: DataMarker<'s>,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Eq,
+{
+}
+
+#[test]
+fn test_clone_eq() {
+    use crate::marker::CowStrMarker;
+    let p1 = DataPayload::<CowStrMarker>::from_borrowed("Demo");
+    let p2 = p1.clone();
+    assert_eq!(p1, p2);
 }
 
 impl<'d, 's, M> DataPayload<'d, 's, M>

--- a/utils/yoke/src/trait_hack.rs
+++ b/utils/yoke/src/trait_hack.rs
@@ -207,12 +207,29 @@ use core::mem;
 
 /// A wrapper around a type `T`, forwarding trait calls down to the inner type.
 ///
-/// `YokeTraitHack` supports [`Clone`] and [`serde::Deserialize`] out of the box. Other traits can
-/// be implemented by the caller.
+/// `YokeTraitHack` supports [`Clone`], [`PartialEq`], [`Eq`], and [`serde::Deserialize`] out of
+/// the box. Other traits can be implemented by the caller.
 ///
 /// For more information, see the module-level documentation.
+///
+/// # Example
+///
+/// Using `YokeTraitHack` as a type bound in a function comparing two `Yoke`s:
+///
+/// ```
+/// use yoke::*;
+/// use yoke::trait_hack::YokeTraitHack;
+///
+/// fn compare_yokes<Y, C1, C2>(y1: Yoke<Y, C1>, y2: Yoke<Y, C2>) -> bool
+/// where
+///     Y: for<'a> Yokeable<'a>,
+///     for<'a> YokeTraitHack<<Y as Yokeable<'a>>::Output>: PartialEq,
+/// {
+///     YokeTraitHack(y1.get()).into_ref() == YokeTraitHack(y2.get()).into_ref()
+/// }
+/// ```
 #[repr(transparent)]
-#[derive(Clone)]
+#[derive(Clone, PartialEq, Eq)]
 pub struct YokeTraitHack<T>(pub T);
 
 impl<'a, T> YokeTraitHack<&'a T> {

